sdk/samples/iot/hub/src/paho_iot_hub_pnp_example.c

```diff
@@ -0,0 +1,1020 @@
+  if ((rc = subscribe()) != 0)
+  {
+    return rc;
+  }
+
+  char* incoming_message_topic;
+  int incoming_message_topic_len;
+  MQTTClient_message* message;
+
+  // First get the twin document to check for updated desired properties. Will then parse desired
+  // property and update accordingly.
+  send_twin_get_message();
+
+  while (device_operational)
+  {
+    // Receive any incoming messages from twin or commands
+    if (MQTTClient_receive(
+            mqtt_client,
+            &incoming_message_topic,
+            &incoming_message_topic_len,
+            &message,
+            TIMEOUT_WAIT_FOR_RECEIVE_MESSAGE_MS)
+            == MQTTCLIENT_SUCCESS
+        && incoming_message_topic != NULL)
+    {
+      on_received(incoming_message_topic, incoming_message_topic_len, message);
+
+      MQTTClient_freeMessage(&message);
+      MQTTClient_free(incoming_message_topic);
+    }
+
+    // Send a telemetry message
+    send_telemetry_message();
+
+    sleep_for_seconds(1);
+  }
+
+  (void)getchar();
+
+  // Gracefully disconnect: send the disconnect packet and close the socket
+  if ((rc = MQTTClient_disconnect(mqtt_client, TIMEOUT_MQTT_DISCONNECT_MS)) != MQTTCLIENT_SUCCESS)
+  {
+    printf("Failed to disconnect MQTT client, return code %d\n", rc);
+    return rc;
+  }
+  printf("Disconnected.\n");
+
+  // Clean up and release resources allocated by the mqtt client
+  MQTTClient_destroy(&mqtt_client);
+
+  return 0;
+}
+
+static void sleep_for_seconds(uint32_t seconds)
+{
+#ifdef _WIN32
+  Sleep((DWORD)seconds * 1000);
+#else
+  sleep(seconds);
+#endif
+}
+
+// Read OS environment variables using stdlib function
+static az_result read_configuration_entry(
+    const char* env_name,
+    char* default_value,
+    bool hide_value,
+    az_span buffer,
+    az_span* out_value)
+{
+  printf("%s = ", env_name);
+  char* env_value = getenv(env_name);
+
+  if (env_value == NULL && default_value != NULL)
+  {
+    env_value = default_value;
+  }
+
+  if (env_value != NULL)
+  {
+    printf("%s\n", hide_value ? "***" : env_value);
+    az_span env_span = az_span_from_str(env_value);
+    AZ_RETURN_IF_NOT_ENOUGH_SIZE(buffer, az_span_size(env_span));
+    az_span_copy(buffer, env_span);
+    *out_value = az_span_slice(buffer, 0, az_span_size(env_span));
+  }
+  else
+  {
+    printf("(missing) Please set the %s environment variable.\n", env_name);
+    return AZ_ERROR_ARG;
+  }
+
+  return AZ_OK;
+}
+
+static int mqtt_publish_message(char* topic, az_span payload, int qos)
+{
+  int rc;
+  MQTTClient_deliveryToken token;
+  if ((rc = MQTTClient_publish(
+           mqtt_client, topic, az_span_size(payload), az_span_ptr(payload), qos, 0, &token))
+      != MQTTCLIENT_SUCCESS)
+  {
+    printf("Unable to publish message, return code %d\n", rc);
+    return rc;
+  }
+  if (qos > 0)
+  {
+    if ((rc = MQTTClient_waitForCompletion(mqtt_client, token, TIMEOUT_WAIT_FOR_COMPLETION_MS))
```

Comment:
>MQTTClient_waitForCompletion [](start = 14, length = 28)

Do we need to do anything with qos!=0 here?  So just straight hard-code to 0 in MQTTCLIENT_publish call?

Even paho_iot_hub_sas_telemetry_example.c whose only purpose in life is to send telemetry just hardcodes to qos0.  And none of callers in this function bother setting it.

If we don't get a PUBACK does that mean that mqtt_publish_message should start retrying at that point?

Reply:
I worry if someone wants to go in and make it a different qos, the behavior could be all wrong. Doing it this way protects against that since there are at least a few asking about making telemetry qos 1.